### PR TITLE
Initial Shrink implementation

### PR DESCRIPTION
**Pull request by nbardiuk**
Hello,

I've prepared a simple implementation of shrink procedure. 
It has a port of scalacheck shrink object for numbers, lists and strings.

This change affects several parts of test library:
  * new shrink object
  * Property stores and perform shrinks
  * CheckResult provides type safe sample and stores number of shrinks 

I understand that it is a big change and probably has a lot of discussion points. We can consider this pull request as a proof of concept and use it to define how shrink functionality will be integrated into the library.

Some explanation how it currently works.

## Shrink
Shrink is a function from value to stream of smaller values
    
    T -> Stream<T>

For example numbers are reduced to 0 and alternate between positive and negative values
    
    shrink(6) = (3, -3, 1, -1, 0)
    shrink(9) = (4, -4, 2, -2, 1, -1, 0)

List shrinks in 2 direction: size and value
     
    shrink([1, 3]) = ([1], [3], [0, 3], [1, 1], [1, -1], [1, 0])

String shrinks as a list of chars.

## Shrinkage
Shrinking process is a recursive function that finds next failure among reduced values. 

    (Shrink, Predicate, CheckResult) -> CheckResult

It exits recursion when all shrinks for last failure satisfy property predicate. The result is last known failure. 

## Property
By default Property has no shrinks configured, that makes shrinkage optional. Developer should explicitly provide shrink objects that match value generators.

```java
Property.def("Is not less")
.forAll(Arbitrary.integer(), Arbitrary.integer())
.suchThat((a, b) -> a >= b)
.shrinking(Shrink.integer(), Shrink.integer())
.check()
.assertIsSatisfied();
```
    Is not less: Falsified after 2 passed tests in 19 ms.
    java.lang.AssertionError: Expected satisfied check result but was 
    Falsified(propertyName = Is not less, count = 3, sample = (0, 1), shrinks = 12)

Without shrinking it looks like before
```java
Property.def("Is not less")
.forAll(Arbitrary.integer(), Arbitrary.integer())
.suchThat((a, b) -> a >= b)
.check()
.assertIsSatisfied();
```
    Is not less: Falsified after 3 passed tests in 11 ms.
    java.lang.AssertionError: Expected satisfied check result but was 
    Falsified(propertyName = Is not less, count = 4, sample = (-34, 62), shrinks = 0)

### CODE REVIEW

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    /**
+     * Generates an arbitrary tuple of 2 given variables
+     *
+     * @param a1 1st variable of this tuple
+     * @param a2 2nd variable of this tuple
+     * @param <T1> 1st variable type of this tuple
+     * @param <T2> 2nd variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2> Arbitrary<Tuple2<T1, T2>> of(Arbitrary<T1> a1, Arbitrary<T2> a2) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+     *
+     * @param a1 1st variable of this tuple
+     * @param a2 2nd variable of this tuple
+     * @param a3 3rd variable of this tuple
+     * @param <T1> 1st variable type of this tuple
+     * @param <T2> 2nd variable type of this tuple
+     * @param <T3> 3rd variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2, T3> Arbitrary<Tuple3<T1, T2, T3>> of(Arbitrary<T1> a1, Arbitrary<T2> a2, Arbitrary<T3> a3) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+     * @param a2 2nd variable of this tuple
+     * @param a3 3rd variable of this tuple
+     * @param a4 4th variable of this tuple
+     * @param <T1> 1st variable type of this tuple
+     * @param <T2> 2nd variable type of this tuple
+     * @param <T3> 3rd variable type of this tuple
+     * @param <T4> 4th variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2, T3, T4> Arbitrary<Tuple4<T1, T2, T3, T4>> of(Arbitrary<T1> a1, Arbitrary<T2> a2, Arbitrary<T3> a3, Arbitrary<T4> a4) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+     * @param a4 4th variable of this tuple
+     * @param a5 5th variable of this tuple
+     * @param <T1> 1st variable type of this tuple
+     * @param <T2> 2nd variable type of this tuple
+     * @param <T3> 3rd variable type of this tuple
+     * @param <T4> 4th variable type of this tuple
+     * @param <T5> 5th variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2, T3, T4, T5> Arbitrary<Tuple5<T1, T2, T3, T4, T5>> of(Arbitrary<T1> a1, Arbitrary<T2> a2, Arbitrary<T3> a3, Arbitrary<T4> a4, Arbitrary<T5> a5) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+     * @param a6 6th variable of this tuple
+     * @param <T1> 1st variable type of this tuple
+     * @param <T2> 2nd variable type of this tuple
+     * @param <T3> 3rd variable type of this tuple
+     * @param <T4> 4th variable type of this tuple
+     * @param <T5> 5th variable type of this tuple
+     * @param <T6> 6th variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2, T3, T4, T5, T6> Arbitrary<Tuple6<T1, T2, T3, T4, T5, T6>> of(Arbitrary<T1> a1, Arbitrary<T2> a2, Arbitrary<T3> a3, Arbitrary<T4> a4, Arbitrary<T5> a5, Arbitrary<T6> a6) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T6' must match pattern '^\[A\-Z\]$'\.

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+     * @param <T1> 1st variable type of this tuple
+     * @param <T2> 2nd variable type of this tuple
+     * @param <T3> 3rd variable type of this tuple
+     * @param <T4> 4th variable type of this tuple
+     * @param <T5> 5th variable type of this tuple
+     * @param <T6> 6th variable type of this tuple
+     * @param <T7> 7th variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2, T3, T4, T5, T6, T7> Arbitrary<Tuple7<T1, T2, T3, T4, T5, T6, T7>> of(Arbitrary<T1> a1, Arbitrary<T2> a2, Arbitrary<T3> a3, Arbitrary<T4> a4, Arbitrary<T5> a5, Arbitrary<T6> a6, Arbitrary<T7> a7) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T6' must match pattern '^\[A\-Z\]$'\.

File: javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+     * @param <T2> 2nd variable type of this tuple
+     * @param <T3> 3rd variable type of this tuple
+     * @param <T4> 4th variable type of this tuple
+     * @param <T5> 5th variable type of this tuple
+     * @param <T6> 6th variable type of this tuple
+     * @param <T7> 7th variable type of this tuple
+     * @param <T8> 8th variable type of this tuple
+     * @return A new generator
+     */
+    public static <T1, T2, T3, T4, T5, T6, T7, T8> Arbitrary<Tuple8<T1, T2, T3, T4, T5, T6, T7, T8>> of(Arbitrary<T1> a1, Arbitrary<T2> a2, Arbitrary<T3> a3, Arbitrary<T4> a4, Arbitrary<T5> a5, Arbitrary<T6> a6, Arbitrary<T7> a7, Arbitrary<T8> a8) {
```
Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T6' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang-test/src-gen/main/java/javaslang/test/ArbitraryTuple.java]: \[Checkstyle\] ERROR: Name 'T7' must match pattern '^\[A\-Z\]$'\.

**Comment by danieldietrich** [comment]
Hi @nbardiuk, this looks great (at the very first sight). Give me some time to read through the code :)

Don't be scared of all the Sputnik review comments. We haven't tweaked all rules, yet. Many Sputnik comments are no real issues (like the generic names etc.).

**Comment by danieldietrich** [comment]
Is shrinking used to find a minimal sample that falsifies the property? I.e. is this change related to #551?

**Comment by nbardiuk** [comment]
Yes, it is the main goal. Issue looks related

**Comment by danieldietrich** [comment]
Did you have seen this solution for the ScalaCheck shrinking problem that Generator constraints may not be respected? https://gist.github.com/davidallsopp/f65d73fea8b5e5165fc3

Don't know if it matters...

**Comment by nbardiuk** [comment]
Yes it matters.

For now there is a workaround - a filter of shrink

    Shrink.integer().filter(i -> i % 2 == 0)

The proper solution could involve change in generator or arbitrary.
For example generator could provide a predicate that checks that a value fits generator ranges.
Or we could encapsulate Gen and Shrink inside Arbitrary and change them simultaneously by manipulating only Arbitrary.

**Comment by danieldietrich** [comment]
I see a 'proper solution' outside of this PR. The API isn't shipped yet, we may change it afterwards.

Without having looked too deep into it: What if an Arbitrary would have an additional method

```java
interface Arbitrary<T> {

    // ...
    
    // or maybe: Shrink<T> shrink();
    Option<Shrink<T>> shrink();
}
```

Given that a 'default' Shrink could be used, unless the user specifies one.

**Comment by nbardiuk** [comment]
Yes, for me it would be the best solution.

But it will require one more change in map function of Arbitrary.
In order to transform Shrink<T> into Shrink<U> we need 2 function T->U and U->T because type parameter of Shrink is in both domain and codomain.
 
By the way we don't need Option<Shrink> there is a Shrink.empty() that has the same meaning.

**Comment by danieldietrich** [comment]
Sounds great. I would suggest to prepare it in a second PR/branch if you don't mind.

**Comment by nbardiuk** [comment]
Sure, I will prepare this change

**Comment by danieldietrich** [comment]
My first, naive thought of solving #551 was to walk into a _shrinking look_ if the first run was a failure and reduce the _size_ parameter step-by-step, until no more failure could be found. That wouldn't require the notion of a 'Shrink'.

What drawbacks do have the solution mentioned above and what can be done with Shrink beyond it?

**Comment by nbardiuk** [comment]
For simple example it sounds the same. 

Maybe composition of Shrink objects is more powerful. For example reducing size for list generator will shrink length of list and its values simultaneously. Current implementation of shrink reduces values and length of list separately.

If smaller example is smaller only in one dimension (ie smaller values but the same length) shrink has better chance to find it.

**Comment by emmanueltouzery** [comment]
noticed this twitter comment on the separation of arbitrary and shrink, maybe it's helpful =>
https://twitter.com/mapastr/status/801070809880166401

**Comment by danieldietrich** [comment]
This PR has been parked for a long time.
We have to look at this topic again _after_ Vavr 1.0.0 release (which will still take some time).

**Pull request closed without merging**